docs/docs/reference/contextual/derivation.md

```diff
@@ -1,362 +1,365 @@
 }
 ```
-would produce:
+
+we need to implement a method `Eq.derived` on the companion object of `Eq` that produces an instance for `Eq[T]` given
+a `Mirror[T]`. Here is a possible implementation,
+
 ```scala
-object Result {
-  import scala.compiletime.Shape._
-
-  given [T, E] as Generic[Result[T, E]] {
-    type Shape = Cases[(
-      Case[Ok[T], T *: Unit],
-      Case[Err[E], E *: Unit]
-    )]
-    ...
+inline given derived[T] as Eq[T] given (m: Mirror.Of[T]) = {
+  val elemInstances = summonAll[m.MirroredElemTypes]           // (1)
+  inline m match {                                             // (2)
+    case s: Mirror.SumOf[T]     => eqSum(s, elemInstances)
+    case p: Mirror.ProductOf[T] => eqProduct(p, elemInstances)
   }
 }
 ```
-The `Generic` class is defined in package `scala.reflect`.
 
-```scala
-abstract class Generic[T] {
-  type Shape <: scala.compiletime.Shape
+Note that the `derived` method is defined as both `inline` and `given`. This means that the method will be expanded at
+call sites (for instance the compiler generated instance definitions in the companion objects of ADTs which have a
+`derived Eq` clause), and also that it can be used recursively if necessary, to compute instance for children.
 
-  /** The mirror corresponding to ADT instance `x` */
-  def reflect(x: T): Mirror
+The body of this method (1) first materializes the `Eq` instances for all the child types of type the instance is
+being derived for. This is either all the branches of a sum type or all the fields of a product type. The
+implementation of `summonAll` is `inline` and uses Dotty's `given match` construct to collect the instances as a
+`List`,
 
-  /** The ADT instance corresponding to given `mirror` */
-  def reify(mirror: Mirror): T
+```scala
+inline def summon[T]: T = given match {
+  case t: T => t
+}
 
-  /** The companion object of the ADT */
-  def common: GenericClass
+inline def summonAll[T <: Tuple]: List[Eq[_]] = inline erasedValue[T] match {
+  case _: Unit => Nil
+  case _: (t *: ts) => summon[Eq[t]] :: summonAll[ts]
 }
 ```
-It defines the `Shape` type for the ADT `T`, as well as two methods that map between a
-type `T` and a generic representation of `T`, which we call a `Mirror`:
-The `reflect` method maps an instance of the ADT `T` to its mirror whereas
-the `reify` method goes the other way. There's also a `common` method that returns
-a value of type `GenericClass` which contains information that is the same for all
-instances of a class (right now, this consists of the runtime `Class` value and
-the names of the cases and their parameters).
 
-### Mirrors
+with the instances for children in hand the `derived` method uses an `inline match` to dispatch to methods which can
+construct instances for either sums or products (2). Note that because `derived` is `inline` the match will be
+resolved at compile-time and only the left-hand side of the matching case will be inlined into the generated code with
```

Comment:
~~~ suggestion
resolved at compile-time and only the right-hand side of the matching case will be inlined into the generated code with
~~~